File: backend/fiscal/src/model/fiscal_parametro_model.py
from src import db
from src.model.fiscal_inscricoes_substitutas_model import FiscalInscricoesSubstitutasModel
from src.model.fiscal_estadual_regime_model import FiscalEstadualRegimeModel
from src.model.fiscal_estadual_porte_model import FiscalEstadualPorteModel
from src.model.fiscal_municipal_regime_model import FiscalMunicipalRegimeModel


class FiscalParametroModel(db.Model):
    __tablename__ = 'fiscal_parametro'

    id = db.Column(db.Integer, primary_key=True)
    vigencia = db.Column(db.String(7))
    descricao_vigencia = db.Column(db.String(100))
    criterio_lancamento = db.Column(db.String(1))
    apuracao = db.Column(db.String(1))
    microempree_individual = db.Column(db.String(1))
    calc_pis_cofins_efd = db.Column(db.String(2))
    simples_codigo_acesso = db.Column(db.String(50))
    simples_tabela = db.Column(db.String(1))
    simples_atividade = db.Column(db.String(2))
    perfil_sped = db.Column(db.String(1))
    apuracao_consolidada = db.Column(db.String(1))
    substituicao_tributaria = db.Column(db.String(1))
    forma_calculo_iss = db.Column(db.String(2))
    id_fiscal_estadual_regime = db.Column(db.Integer, db.ForeignKey('fiscal_estadual_regime.id'))
    id_fiscal_estadual_porte = db.Column(db.Integer, db.ForeignKey('fiscal_estadual_porte.id'))
    id_fiscal_municipal_regime = db.Column(db.Integer, db.ForeignKey('fiscal_municipal_regime.id'))

    fiscal_inscricoes_substitutas_model_list = db.relationship('FiscalInscricoesSubstitutasModel', lazy='dynamic')
    fiscal_estadual_regime_model = db.relationship('FiscalEstadualRegimeModel', foreign_keys=[id_fiscal_estadual_regime])
    fiscal_estadual_porte_model = db.relationship('FiscalEstadualPorteModel', foreign_keys=[id_fiscal_estadual_porte])
    fiscal_municipal_regime_model = db.relationship('FiscalMunicipalRegimeModel', foreign_keys=[id_fiscal_municipal_regime])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_fiscal_estadual_porte = data.get('idFiscalEstadualPorte')
        self.id_fiscal_estadual_regime = data.get('idFiscalEstadualRegime')
        self.id_fiscal_municipal_regime = data.get('idFiscalMunicipalRegime')
        self.vigencia = data.get('vigencia')
        self.descricao_vigencia = data.get('descricaoVigencia')
        self.criterio_lancamento = data.get('criterioLancamento')
        self.apuracao = data.get('apuracao')
        self.microempree_individual = data.get('microempreeIndividual')
        self.calc_pis_cofins_efd = data.get('calcPisCofinsEfd')
        self.simples_codigo_acesso = data.get('simplesCodigoAcesso')
        self.simples_tabela = data.get('simplesTabela')
        self.simples_atividade = data.get('simplesAtividade')
        self.perfil_sped = data.get('perfilSped')
        self.apuracao_consolidada = data.get('apuracaoConsolidada')
        self.substituicao_tributaria = data.get('substituicaoTributaria')
        self.forma_calculo_iss = data.get('formaCalculoIss')

    def serialize(self):
        return {
            'id': self.id,
            'idFiscalEstadualPorte': self.id_fiscal_estadual_porte,
            'idFiscalEstadualRegime': self.id_fiscal_estadual_regime,
            'idFiscalMunicipalRegime': self.id_fiscal_municipal_regime,
            'vigencia': self.vigencia,
            'descricaoVigencia': self.descricao_vigencia,
            'criterioLancamento': self.criterio_lancamento,
            'apuracao': self.apuracao,
            'microempreeIndividual': self.microempree_individual,
            'calcPisCofinsEfd': self.calc_pis_cofins_efd,
            'simplesCodigoAcesso': self.simples_codigo_acesso,
            'simplesTabela': self.simples_tabela,
            'simplesAtividade': self.simples_atividade,
            'perfilSped': self.perfil_sped,
            'apuracaoConsolidada': self.apuracao_consolidada,
            'substituicaoTributaria': self.substituicao_tributaria,
            'formaCalculoIss': self.forma_calculo_iss,
            'fiscalInscricoesSubstitutasModelList': [fiscal_inscricoes_substitutas_model.serialize() for fiscal_inscricoes_substitutas_model in self.fiscal_inscricoes_substitutas_model_list],
            'fiscalEstadualRegimeModel': self.fiscal_estadual_regime_model.serialize() if self.fiscal_estadual_regime_model else None,
            'fiscalEstadualPorteModel': self.fiscal_estadual_porte_model.serialize() if self.fiscal_estadual_porte_model else None,
            'fiscalMunicipalRegimeModel': self.fiscal_municipal_regime_model.serialize() if self.fiscal_municipal_regime_model else None,
        }

File: frontend/fiscal/lib/app/data/model/fiscal_parametro_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:fiscal/app/infra/infra_imports.dart';
import 'package:fiscal/app/data/model/model_imports.dart';
import 'package:fiscal/app/data/domain/domain_imports.dart';

class FiscalParametroModel {
	int? id;
	int? idFiscalEstadualPorte;
	int? idFiscalEstadualRegime;
	int? idFiscalMunicipalRegime;
	String? vigencia;
	String? descricaoVigencia;
	String? criterioLancamento;
	String? apuracao;
	String? microempreeIndividual;
	String? calcPisCofinsEfd;
	String? simplesCodigoAcesso;
	String? simplesTabela;
	String? simplesAtividade;
	String? perfilSped;
	String? apuracaoConsolidada;
	String? substituicaoTributaria;
	String? formaCalculoIss;
	List<FiscalInscricoesSubstitutasModel>? fiscalInscricoesSubstitutasModelList;
	FiscalEstadualRegimeModel? fiscalEstadualRegimeModel;
	FiscalEstadualPorteModel? fiscalEstadualPorteModel;
	FiscalMunicipalRegimeModel? fiscalMunicipalRegimeModel;

	FiscalParametroModel({
		this.id,
		this.idFiscalEstadualPorte,
		this.idFiscalEstadualRegime,
		this.idFiscalMunicipalRegime,
		this.vigencia,
		this.descricaoVigencia,
		this.criterioLancamento,
		this.apuracao,
		this.microempreeIndividual,
		this.calcPisCofinsEfd,
		this.simplesCodigoAcesso,
		this.simplesTabela,
		this.simplesAtividade,
		this.perfilSped,
		this.apuracaoConsolidada,
		this.substituicaoTributaria,
		this.formaCalculoIss,
		this.fiscalInscricoesSubstitutasModelList,
		this.fiscalEstadualRegimeModel,
		this.fiscalEstadualPorteModel,
		this.fiscalMunicipalRegimeModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'vigencia',
		'descricao_vigencia',
		'criterio_lancamento',
		'apuracao',
		'microempree_individual',
		'calc_pis_cofins_efd',
		'simples_codigo_acesso',
		'simples_tabela',
		'simples_atividade',
		'perfil_sped',
		'apuracao_consolidada',
		'substituicao_tributaria',
		'forma_calculo_iss',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Vigencia',
		'Descricao Vigencia',
		'Criterio Lancamento',
		'Apuracao',
		'Microempree Individual',
		'Calc Pis Cofins Efd',
		'Simples Codigo Acesso',
		'Simples Tabela',
		'Simples Atividade',
		'Perfil Sped',
		'Apuracao Consolidada',
		'Substituicao Tributaria',
		'Forma Calculo Iss',
	];

	FiscalParametroModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idFiscalEstadualPorte = jsonData['idFiscalEstadualPorte'];
		idFiscalEstadualRegime = jsonData['idFiscalEstadualRegime'];
		idFiscalMunicipalRegime = jsonData['idFiscalMunicipalRegime'];
		vigencia = jsonData['vigencia'];
		descricaoVigencia = jsonData['descricaoVigencia'];
		criterioLancamento = FiscalParametroDomain.getCriterioLancamento(jsonData['criterioLancamento']);
		apuracao = FiscalParametroDomain.getApuracao(jsonData['apuracao']);
		microempreeIndividual = FiscalParametroDomain.getMicroempreeIndividual(jsonData['microempreeIndividual']);
		calcPisCofinsEfd = FiscalParametroDomain.getCalcPisCofinsEfd(jsonData['calcPisCofinsEfd']);
		simplesCodigoAcesso = jsonData['simplesCodigoAcesso'];
		simplesTabela = FiscalParametroDomain.getSimplesTabela(jsonData['simplesTabela']);
		simplesAtividade = FiscalParametroDomain.getSimplesAtividade(jsonData['simplesAtividade']);
		perfilSped = FiscalParametroDomain.getPerfilSped(jsonData['perfilSped']);
		apuracaoConsolidada = FiscalParametroDomain.getApuracaoConsolidada(jsonData['apuracaoConsolidada']);
		substituicaoTributaria = FiscalParametroDomain.getSubstituicaoTributaria(jsonData['substituicaoTributaria']);
		formaCalculoIss = FiscalParametroDomain.getFormaCalculoIss(jsonData['formaCalculoIss']);
		fiscalInscricoesSubstitutasModelList = (jsonData['fiscalInscricoesSubstitutasModelList'] as Iterable?)?.map((m) => FiscalInscricoesSubstitutasModel.fromJson(m)).toList() ?? [];
		fiscalEstadualRegimeModel = jsonData['fiscalEstadualRegimeModel'] == null ? FiscalEstadualRegimeModel() : FiscalEstadualRegimeModel.fromJson(jsonData['fiscalEstadualRegimeModel']);
		fiscalEstadualPorteModel = jsonData['fiscalEstadualPorteModel'] == null ? FiscalEstadualPorteModel() : FiscalEstadualPorteModel.fromJson(jsonData['fiscalEstadualPorteModel']);
		fiscalMunicipalRegimeModel = jsonData['fiscalMunicipalRegimeModel'] == null ? FiscalMunicipalRegimeModel() : FiscalMunicipalRegimeModel.fromJson(jsonData['fiscalMunicipalRegimeModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idFiscalEstadualPorte'] = idFiscalEstadualPorte != 0 ? idFiscalEstadualPorte : null;
		jsonData['idFiscalEstadualRegime'] = idFiscalEstadualRegime != 0 ? idFiscalEstadualRegime : null;
		jsonData['idFiscalMunicipalRegime'] = idFiscalMunicipalRegime != 0 ? idFiscalMunicipalRegime : null;
		jsonData['vigencia'] = Util.removeMask(vigencia);
		jsonData['descricaoVigencia'] = descricaoVigencia;
		jsonData['criterioLancamento'] = FiscalParametroDomain.setCriterioLancamento(criterioLancamento);
		jsonData['apuracao'] = FiscalParametroDomain.setApuracao(apuracao);
		jsonData['microempreeIndividual'] = FiscalParametroDomain.setMicroempreeIndividual(microempreeIndividual);
		jsonData['calcPisCofinsEfd'] = FiscalParametroDomain.setCalcPisCofinsEfd(calcPisCofinsEfd);
		jsonData['simplesCodigoAcesso'] = simplesCodigoAcesso;
		jsonData['simplesTabela'] = FiscalParametroDomain.setSimplesTabela(simplesTabela);
		jsonData['simplesAtividade'] = FiscalParametroDomain.setSimplesAtividade(simplesAtividade);
		jsonData['perfilSped'] = FiscalParametroDomain.setPerfilSped(perfilSped);
		jsonData['apuracaoConsolidada'] = FiscalParametroDomain.setApuracaoConsolidada(apuracaoConsolidada);
		jsonData['substituicaoTributaria'] = FiscalParametroDomain.setSubstituicaoTributaria(substituicaoTributaria);
		jsonData['formaCalculoIss'] = FiscalParametroDomain.setFormaCalculoIss(formaCalculoIss);
		
		var fiscalInscricoesSubstitutasModelLocalList = []; 
		for (FiscalInscricoesSubstitutasModel object in fiscalInscricoesSubstitutasModelList ?? []) { 
			fiscalInscricoesSubstitutasModelLocalList.add(object.toJson); 
		}
		jsonData['fiscalInscricoesSubstitutasModelList'] = fiscalInscricoesSubstitutasModelLocalList;
		jsonData['fiscalEstadualRegimeModel'] = fiscalEstadualRegimeModel?.toJson;
		jsonData['fiscalEstadualPorteModel'] = fiscalEstadualPorteModel?.toJson;
		jsonData['fiscalMunicipalRegimeModel'] = fiscalMunicipalRegimeModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idFiscalEstadualPorte = plutoRow.cells['idFiscalEstadualPorte']?.value;
		idFiscalEstadualRegime = plutoRow.cells['idFiscalEstadualRegime']?.value;
		idFiscalMunicipalRegime = plutoRow.cells['idFiscalMunicipalRegime']?.value;
		vigencia = plutoRow.cells['vigencia']?.value;
		descricaoVigencia = plutoRow.cells['descricaoVigencia']?.value;
		criterioLancamento = plutoRow.cells['criterioLancamento']?.value != '' ? plutoRow.cells['criterioLancamento']?.value : 'Livre';
		apuracao = plutoRow.cells['apuracao']?.value != '' ? plutoRow.cells['apuracao']?.value : '1-Regime Competencia';
		microempreeIndividual = plutoRow.cells['microempreeIndividual']?.value != '' ? plutoRow.cells['microempreeIndividual']?.value : 'Sim';
		calcPisCofinsEfd = plutoRow.cells['calcPisCofinsEfd']?.value != '' ? plutoRow.cells['calcPisCofinsEfd']?.value : 'AB=Alíquota Básica';
		simplesCodigoAcesso = plutoRow.cells['simplesCodigoAcesso']?.value;
		simplesTabela = plutoRow.cells['simplesTabela']?.value != '' ? plutoRow.cells['simplesTabela']?.value : '1=Federal';
		simplesAtividade = plutoRow.cells['simplesAtividade']?.value != '' ? plutoRow.cells['simplesAtividade']?.value : 'Comercio';
		perfilSped = plutoRow.cells['perfilSped']?.value != '' ? plutoRow.cells['perfilSped']?.value : 'A';
		apuracaoConsolidada = plutoRow.cells['apuracaoConsolidada']?.value != '' ? plutoRow.cells['apuracaoConsolidada']?.value : 'Sim';
		substituicaoTributaria = plutoRow.cells['substituicaoTributaria']?.value != '' ? plutoRow.cells['substituicaoTributaria']?.value : 'Sim';
		formaCalculoIss = plutoRow.cells['formaCalculoIss']?.value != '' ? plutoRow.cells['formaCalculoIss']?.value : 'Normal';
		fiscalInscricoesSubstitutasModelList = [];
		fiscalEstadualRegimeModel = FiscalEstadualRegimeModel();
		fiscalEstadualRegimeModel?.nome = plutoRow.cells['fiscalEstadualRegimeModel']?.value;
		fiscalEstadualPorteModel = FiscalEstadualPorteModel();
		fiscalEstadualPorteModel?.nome = plutoRow.cells['fiscalEstadualPorteModel']?.value;
		fiscalMunicipalRegimeModel = FiscalMunicipalRegimeModel();
		fiscalMunicipalRegimeModel?.nome = plutoRow.cells['fiscalMunicipalRegimeModel']?.value;
	}	

	FiscalParametroModel clone() {
		return FiscalParametroModel(
			id: id,
			idFiscalEstadualPorte: idFiscalEstadualPorte,
			idFiscalEstadualRegime: idFiscalEstadualRegime,
			idFiscalMunicipalRegime: idFiscalMunicipalRegime,
			vigencia: vigencia,
			descricaoVigencia: descricaoVigencia,
			criterioLancamento: criterioLancamento,
			apuracao: apuracao,
			microempreeIndividual: microempreeIndividual,
			calcPisCofinsEfd: calcPisCofinsEfd,
			simplesCodigoAcesso: simplesCodigoAcesso,
			simplesTabela: simplesTabela,
			simplesAtividade: simplesAtividade,
			perfilSped: perfilSped,
			apuracaoConsolidada: apuracaoConsolidada,
			substituicaoTributaria: substituicaoTributaria,
			formaCalculoIss: formaCalculoIss,
			fiscalInscricoesSubstitutasModelList: fiscalInscricoesSubstitutasModelListClone(fiscalInscricoesSubstitutasModelList!),
		);			
	}

	fiscalInscricoesSubstitutasModelListClone(List<FiscalInscricoesSubstitutasModel> fiscalInscricoesSubstitutasModelList) { 
		List<FiscalInscricoesSubstitutasModel> resultList = [];
		for (var fiscalInscricoesSubstitutasModel in fiscalInscricoesSubstitutasModelList) {
			resultList.add(
				FiscalInscricoesSubstitutasModel(
					id: fiscalInscricoesSubstitutasModel.id,
					idFiscalParametros: fiscalInscricoesSubstitutasModel.idFiscalParametros,
					uf: fiscalInscricoesSubstitutasModel.uf,
					inscricaoEstadual: fiscalInscricoesSubstitutasModel.inscricaoEstadual,
					pmpf: fiscalInscricoesSubstitutasModel.pmpf,
				)
			);
		}
		return resultList;
	}

	
}